psi-sample/client/src/main/java/com/r3/conclave/sample/client/Client.java

```diff
@@ -2,130 +2,148 @@
+
+    /**
+     * Use Kryo to serialize inputs from client to enclave
+     */
+    private Output serializeMessage(InputData listOfCreditCardNumbers){
+        Kryo kryo = new Kryo();
+        Output output = new Output(new ByteArrayOutputStream());
+        kryo.register(InputData.class, new InputDataSerializer());
+        kryo.writeObject(output, listOfCreditCardNumbers);
+        output.close();
+        return output;
+    }
+
+    @Override
+    public Void call() throws Exception {
+        EnclaveClient enclaveClient;
+        if (Files.exists(file)) {
+            //exiting private kay is loaded from the previous saved state.
+            enclaveClient = new EnclaveClient(Files.readAllBytes(file));
+            System.out.println("Loaded previous client state and thus using existing private key.");
+        } else {
+            //a new private key is generated. Enclave Client is created using this private key and constraint.
+            //a corresponding public key will be used by the enclave to encrypt data to be sent to this client
+            System.out.println("No previous client state. Generating new state with new private key.");
+            enclaveClient = new EnclaveClient(constraint);
         }
-        String clientType = args[0];
-        String constraint = args[1];
-
-        //connect to host server
-        DataInputStream fromHost;
-        DataOutputStream toHost;
-        while (true) {
-            try {
-                System.out.println("Attempting to connect to localhost:9999");
-                Socket socket = new Socket();
-                socket.connect(new InetSocketAddress(InetAddress.getLoopbackAddress(), 9999), 5000);
-                fromHost = new DataInputStream(socket.getInputStream());
-                toHost = new DataOutputStream(socket.getOutputStream());
-                break;
-            } catch (Exception e) {
-                System.err.println("Retrying: " + e.getMessage());
-                Thread.sleep(2000);
+
+        try (WebEnclaveTransport transport = new WebEnclaveTransport(url);
+             EnclaveClient client = enclaveClient)
+        {
+            //retrieve the enclaveInstanceInfo object, i.e. the remote attestation object and verify it against the
+            //constraint
+            client.start(transport);
+
+            //collect merchants and service providers credit card numbers list
+            InputData inputData = getInputData();
+
+            byte[] requestMailBody = serializeMessage(inputData).getBuffer();
+
+            //client will send its credit card list to enclave and wait for other client to send their list
+            EnclaveMail responseMail = client.sendMail(requestMailBody);
+
+            //responseMail is null till enclave doesn't reply back to the client
+            if(responseMail == null) {
+                do
+                {
+                    //poll for reply to enclave
+                    responseMail = enclaveClient.pollMail();
```

Comment:
You need to add a call to `Thread.sleep`, otherwise the client will be spamming the web server. 5-10 sec delay should be fine.

Reply:
done

Reply:
Minor point, but it will be better to have the sleep before the `pollMail`. Because if `pollMail` did actually return a Mail object then you still have to wait 10s to get it!

Reply:
done